File: installer/manifest.py
from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class Manifest:
    data: dict[str, Any]
    source: Path

    def __getitem__(self, key: str) -> Any:
        return self.data[key]


def load_manifest(path: Path) -> Manifest:
    path = Path(path)
    return Manifest(json.loads(path.read_text(encoding='utf-8')), path)


def _frontmatter_name(skill_path: Path) -> str | None:
    text = skill_path.read_text(encoding='utf-8')
    if not text.startswith('---\n'):
        return None
    end = text.find('\n---\n', 4)
    if end == -1:
        return None
    for line in text[4:end].splitlines():
        if line.startswith('name:'):
            return line.split(':', 1)[1].strip().strip('"\'')
    return None


def validate_manifest(manifest: Manifest, package_root: Path) -> list[str]:
    root = Path(package_root).resolve()
    data = manifest.data
    errors: list[str] = []

    skill = data.get('skill') or {}
    entry = skill.get('entrypoint')
    name = skill.get('name')
    if entry != 'SKILL.md':
        errors.append('Canonical entrypoint must be SKILL.md')

    skill_path = root / 'SKILL.md'
    if not skill_path.is_file():
        errors.append('Missing canonical SKILL.md at package root')
    else:
        fm_name = _frontmatter_name(skill_path)
        if fm_name != name:
            errors.append(f'Manifest skill name {name!r} does not match SKILL.md name {fm_name!r}')

    payload = data.get('payload')
    if not isinstance(payload, list) or not payload:
        errors.append('Payload must be a non-empty list')
        payload = []

    seen: set[str] = set()
    for raw in payload:
        if not isinstance(raw, str) or not raw:
            errors.append('Payload entries must be non-empty strings')
            continue
        if raw in seen:
            errors.append(f'Duplicate payload entry: {raw}')
            continue
        seen.add(raw)
        rel = Path(raw)
        if rel.is_absolute() or '..' in rel.parts:
            errors.append(f'Payload path traversal/outside package root: {raw}')
            continue
        resolved = (root / rel).resolve()
        try:
            resolved.relative_to(root)
        except ValueError:
            errors.append(f'Payload path outside package root: {raw}')
            continue
        if not resolved.exists():
            errors.append(f'Payload entry does not exist: {raw}')

    prerequisites = data.get('prerequisites', {})
    if not isinstance(prerequisites, dict):
        errors.append('Prerequisites must be an object')
    elif 'superpowers' in prerequisites:
        errors.append('Standalone manifest must not require an external process framework')

    profiles = data.get('document_profiles')
    if not isinstance(profiles, dict) or set(profiles) != {'standalone', 'oai-native'}:
        errors.append('Document profiles must be exactly standalone and oai-native')

    if 'skills' not in payload:
        errors.append('Standalone runtime payload must include package-owned skills')

    return errors

File: installer/node/manifest.js
import { createHash } from "node:crypto";
import {
  accessSync,
  constants,
  existsSync,
  lstatSync,
  readlinkSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync
} from "node:fs";
import path from "node:path";

export class InstallerError extends Error {}

const EXCLUDED_DIRECTORIES = new Set(["__pycache__", ".pytest_cache"]);
const WINDOWS_DRIVE = /^[A-Za-z]:/;

export function physicalPath(candidate, links = 0) {
  const absolute = path.resolve(candidate);
  if (links > 40) throw new InstallerError(`too many symbolic links: ${absolute}`);
  try {
    if (lstatSync(absolute).isSymbolicLink()) {
      // Resolve dangling aliases too: the target can temporarily be absent
      // between backup and replacement while its sibling lock remains held.
      return physicalPath(path.resolve(path.dirname(absolute), readlinkSync(absolute)), links + 1);
    }
    return realpathSync.native(absolute);
  } catch (error) {
    if (error.code !== 'ENOENT') throw error;
    const parent = path.dirname(absolute);
    if (parent === absolute) return absolute;
    return path.join(physicalPath(parent, links), path.basename(absolute));
  }
}

function isInside(root, candidate) {
  const relative = path.relative(root, candidate);
  return relative === "" || (
    relative !== ".." &&
    !relative.startsWith(`..${path.sep}`) &&
    !path.isAbsolute(relative)
  );
}

export function managedPath(root, raw) {
  if (typeof raw !== "string" || raw.trim() === "")
    throw new InstallerError("managed payload path must be a non-empty relative path");
  const normalized = raw.replaceAll("\\", "/");
  const parts = normalized.split("/");
  if (
    normalized.startsWith("/") ||
    normalized.startsWith("//") ||
    WINDOWS_DRIVE.test(normalized) ||
    parts.some((part) => part === "" || part === "." || part === "..")
  ) throw new InstallerError(`managed payload path escapes install root: ${JSON.stringify(raw)}`);
  const physicalRoot = physicalPath(root);
  const candidate = physicalPath(path.join(root, ...parts));
  if (!isInside(physicalRoot, candidate))
    throw new InstallerError(`managed payload path escapes install root: ${JSON.stringify(raw)}`);
  return candidate;
}

export function loadManifest(manifestPath) {
  return JSON.parse(readFileSync(manifestPath, "utf8"));
}

function frontmatterName(skillPath) {
  const text = readFileSync(skillPath, "utf8");
  const match = text.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n/);
  if (!match) return null;
  const name = match[1].split(/\r?\n/).find((line) => line.startsWith("name:"));
  return name ? name.slice(5).trim().replace(/^["']|["']$/g, "") : null;
}

export function validateManifest(manifest, packageRoot) {
  const errors = [];
  const root = physicalPath(packageRoot);
  if (manifest?.skill?.entrypoint !== "SKILL.md")
    errors.push("Canonical entrypoint must be SKILL.md");
  const skillPath = path.join(root, "SKILL.md");
  if (!existsSync(skillPath)) errors.push("Missing canonical SKILL.md at package root");
  else if (frontmatterName(skillPath) !== manifest?.skill?.name)
    errors.push("Manifest skill name does not match SKILL.md name");
  const payload = manifest?.payload;
  if (!Array.isArray(payload) || payload.length === 0)
    errors.push("Payload must be a non-empty list");
  const seen = new Set();
  for (const raw of Array.isArray(payload) ? payload : []) {
    if (typeof raw !== "string" || raw === "") {
      errors.push("Payload entries must be non-empty strings");
      continue;
    }
    if (seen.has(raw)) errors.push(`Duplicate payload entry: ${raw}`);
    seen.add(raw);
    try {
      const resolved = managedPath(root, raw);
      if (!existsSync(resolved)) errors.push(`Payload entry does not exist: ${raw}`);
    } catch (error) {
      errors.push(error.message);
    }
  }
  if (manifest?.prerequisites?.superpowers)
    errors.push("Standalone manifest must not require an external process framework");
  if (JSON.stringify(Object.keys(manifest?.document_profiles ?? {}).sort()) !== JSON.stringify(["oai-native", "standalone"]))
    errors.push("Document profiles must be exactly standalone and oai-native");
  if (!Array.isArray(payload) || !payload.includes("skills"))
    errors.push("Standalone runtime payload must include package-owned skills");
  return errors;
}

function walk(directory) {
  const files = [];
  for (const entry of readdirSync(directory, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name))) {
    if (EXCLUDED_DIRECTORIES.has(entry.name)) continue;
    const candidate = path.join(directory, entry.name);
    if (entry.isDirectory()) files.push(...walk(candidate));
    else if (entry.isFile()) files.push(candidate);
  }
  return files;
}

export function payloadFiles(packageRoot, manifest) {
  const root = physicalPath(packageRoot);
  const files = [];
  for (const raw of manifest.payload) {
    const candidate = managedPath(root, raw);
    if (statSync(candidate).isDirectory()) files.push(...walk(candidate));
    else files.push(candidate);
  }
  return [...new Set(files)].sort((a, b) => a.localeCompare(b));
}

export function sha256File(filePath) {
  accessSync(filePath, constants.R_OK);
  return createHash("sha256").update(readFileSync(filePath)).digest("hex");
}

export function payloadHashes(packageRoot, manifest) {
  const root = physicalPath(packageRoot);
  return Object.fromEntries(payloadFiles(root, manifest).map((file) => [
    path.relative(root, file).split(path.sep).join("/"),
    sha256File(file)
  ]));
}
